internal/api/attestationconfigapi/cli/client/reporter.go

```diff
@@ -0,0 +1,358 @@
+			}
+		},
+		func() {
+			getMinimalVersion = func() (any, string, error) {
+				return findMinimalVersion[attestationconfigapi.SEVSNPVersion](ctx, attestationVariant, versionDates, c.s3Client, c.cacheWindowSize)
+			}
+		},
+	)
+	return getMinimalVersion()
+}
+
+func findMinimalVersion[T attestationconfigapi.TDXVersion | attestationconfigapi.SEVSNPVersion](
+	ctx context.Context, variant variant.Variant, versionDates []string,
+	s3Client *client.Client, cacheWindowSize int,
+) (T, string, error) {
+	var minimalVersion *T
+	var minimalDate string
+	sort.Sort(sort.Reverse(sort.StringSlice(versionDates))) // sort in reverse order to slice the latest versions
+	versionDates = versionDates[:cacheWindowSize]
+	sort.Strings(versionDates) // sort with oldest first to to take the minimal version with the oldest date
+
+	for _, date := range versionDates {
+		obj, err := client.Fetch(ctx, s3Client, apiVersionObject{version: date + ".json", variant: variant, cached: true})
+		if err != nil {
+			return *new(T), "", fmt.Errorf("get object: %w", err)
+		}
+		obj.variant = variant // variant is not set by Fetch, set it manually
+
+		if minimalVersion == nil {
+			v := obj.getVersion().(T)
+			minimalVersion = &v
+			minimalDate = date
+			continue
+		}
+
+		// If the current minimal version has newer versions than the one we just fetched,
+		// update the minimal version to the older version.
+		if isInputNewerThanOtherVersion(variant, *minimalVersion, obj.getVersion()) {
+			v := obj.getVersion().(T)
+			minimalVersion = &v
+			minimalDate = date
+		}
+	}
+
+	return *minimalVersion, minimalDate, nil
+}
+
+func isInputNewerThanOtherVersion(variant variant.Variant, inputVersion, otherVersion any) bool {
+	var result bool
+	actionForVariant(variant,
+		func() {
+			input := inputVersion.(attestationconfigapi.TDXVersion)
+			other := otherVersion.(attestationconfigapi.TDXVersion)
+			result = isInputNewerThanOtherTDXVersion(input, other)
+		},
+		func() {
+			input := inputVersion.(attestationconfigapi.SEVSNPVersion)
+			other := otherVersion.(attestationconfigapi.SEVSNPVersion)
+			result = isInputNewerThanOtherSEVSNPVersion(input, other)
+		},
+	)
+	return result
+}
+
+type apiVersionObject struct {
+	version string          `json:"-"`
+	variant variant.Variant `json:"-"`
+	cached  bool            `json:"-"`
+	snp     attestationconfigapi.SEVSNPVersion
+	tdx     attestationconfigapi.TDXVersion
+}
+
+func (a apiVersionObject) MarshalJSON() ([]byte, error) {
+	var res []byte
+	var err error
+	actionForVariant(a.variant,
+		func() {
+			res, err = json.Marshal(a.tdx)
+		},
+		func() {
+			res, err = json.Marshal(a.snp)
+		},
+	)
+	return res, err
+}
+
+func (a *apiVersionObject) UnmarshalJSON(data []byte) error {
+	errTDX := json.Unmarshal(data, &a.tdx)
+	errSNP := json.Unmarshal(data, &a.snp)
+	if errTDX == nil || errSNP == nil {
```

Comment:
Should this be XOR?

Reply:
For our use case it does not matter if both versions could be unmarshalled.
It does make this a bit more forwardcompatible though, if we ever decided to include SEV and TDX versions in the same object in s3.